job/watcher.go

```diff
@@ -33,72 +34,105 @@ func NewWatcher(client *kubernetes.Clientset, container string) *Watcher {
 }
 
 // Watch gets pods and tail the logs.
-// At first, finds pods from the job definition, and waits to start the pods.
-// Next, gets log requests and get the output with stream.
+// We must create endless loop because sometimes jobs are configured restartPolicy.
+// When restartPolicy is Never, the Job create a new Pod if the specified command is failed.
+// So we must trace all Pods even though the Pod is failed.
+// And it isn't necessary to stop the loop because the Job is watched in WaitJobComplete.
 func (w *Watcher) Watch(job *v1.Job, ctx context.Context) error {
-	podList, err := w.WaitToStartPods(job)
-	if err != nil {
-		return err
+	currentPodList := []corev1.Pod{}
+retry:
+	for {
+		newPodList, err := w.FindPods(job)
+		if err != nil {
+			return err
+		}
+
+		incrementalPodList := diffPods(currentPodList, newPodList)
+		go w.WatchPods(incrementalPodList, ctx)
+
+		time.Sleep(1 * time.Second)
+		currentPodList = newPodList
+		continue retry
 	}
+	return nil
+}
 
+// WatchPods gets wait to start pod and tail the logs.
+func (w *Watcher) WatchPods(pods []corev1.Pod, ctx context.Context) error {
```

Comment:
__[Golint]__ context.Context should be the first parameter of a function ([link](https://sider.review/gh/repos/169710725/pulls/15#issue-3828088))

<sub>posted by Sider</sub>